cirq/protocols/resolve_parameters.py

```diff
@@ -154,7 +156,17 @@ def resolve_parameters(
         return type(val)(resolve_parameters(e, param_resolver, recursive) for e in val)
 
     getter = getattr(val, '_resolve_parameters_', None)
-    result = NotImplemented if getter is None else getter(param_resolver, recursive)
+    try:
+        result = NotImplemented if getter is None else getter(param_resolver, recursive)
+    except TypeError:
```

Comment:
Is this the only way we can get this exception?  I am not sure this is the best way to catch the function not having recursive.  How about something like this instead?

```
if 'recursive' not in getter.__code__.co_varnames:
   if not recursive:
            raise ValueError(
                f'Object type {type(val)} does not support non-recursive parameter resolution.'
                ' This must be updated before Cirq v0.11.'
            )
     result = NotImplemented if getter is None else getter(param_resolver)
```

I am not sure how version compatible this is, but I think something like this would work.

Reply:
I'll try this first, seeing as it sidesteps the more fragile `try/except` logic.